### PR TITLE
Support non-extension wheels with binary dependencies

### LINKED ISSUE #32: check for binaries in purelib
To help in situations like #28, it would be nice if auditwheel could check for impure files in purelib and complain if they're found.

### LINKED ISSUE #107: Regression in tests from #95 - numpy 1.11.0 is no longer matched to manylinux
The external libraries [showing up in this map with "null" keypairs](https://travis-ci.org/pypa/auditwheel/jobs/399795091#L691-L712) is causing legitimate wheels to be misidentified as non-manylinux. We missed this because the tests appeared to be passing, due to #105.

### PR DESCRIPTION
**Pull request by ehashman**
Fixes #107 (and hence #95).
Also fixes #32.

Behold, a masterpiece. Had to base this on #106 so Travis would work properly.

### CODE REVIEW

File: auditwheel/wheeltools.py [hunk 1]
```diff
@@ -203,11 +202,15 @@ def add_platforms(wheel_ctx, platforms, remove_platforms=()):
...
-    for platform in platforms:
-        if platform not in fname_tags:
-            fname_tags.append(platform)
+    fname_tags = {tag for tag in original_fname_tags
+                  if tag not in remove_platforms}
+    fname_tags |= set(platforms)
+
+    # Can't be 'any' and another platform
+    if 'any' in fname_tags and len(fname_tags) > 1:
+        fname_tags.remove('any')
```
Comment (ehashman) [inline auditwheel/wheeltools.py]: Oh I need to remove 'any' from platforms as well, one moment...

**Comment by safijari** [comment]
Hurraaaaay. I vote we get this and #106 in so #92 or #104 can be merged :D

**Comment by safijari** [comment]
bump

**Comment by ehashman** [comment]
I'm going to add a test for #47 to ensure we're not breaking that logic here before this can be merged.

Since CI is not actually passing right now I will add that test to this PR.

File: auditwheel/wheel_abi.py [hunk 2]
```diff
@@ -43,19 +45,32 @@ def get_wheel_elfdata(wheel_fn: str):
...
-
+                full_external_refs[fn] = lddtree_external_references(elftree,
+                                                                     ctx.path)
+            else:
+                nonpy_elftree[fn] = elftree
+
+        needed_libs = [elf['needed'] for elf
+                       in itertools.chain(full_elftree.values(),
+                                          nonpy_elftree.values())]
+        uniq_needs = set(sum(needed_libs, []))  # concatenate needed libs together, get uniq
```
[the remark this answers is not in the document]
Comment (ehashman) [reply auditwheel/wheel_abi.py]: Sick, I love it

**Comment by ehashman** [comment]
Okay, when CI passes this should be good to go.

**Comment by di** [comment]
I reviewed this along with @ehashman, and while I'm not an `auditwheel` contributor or maintainer, she has thoroughly studied this issue and taken great lengths (including fixing broken tests and adding new tests) to ensure that this change is valid. I'm also confident that she'll follow through on any problems that may arise as a result of this change, so I'm 👍 on merging.

**Comment by ehashman** [comment]
Thanks @di! I'm going to merge this PR to unbreak the tests on master.

**Comment by louisabraham** [comment]
Hi, what is the behavior supposed to be when platform-specific libraries are found in purelib? I get an error message

```
RuntimeError: Invalid binary wheel, found the following shared library/libraries in purelib folder:
	libdivsufsort.so.3.0.1
	libdivsufsort64.so.3.0.1
The wheel has to be platlib compliant in order to be repaired by auditwheel.
```

https://travis-ci.org/louisabraham/pydivsufsort/jobs/657796428

**Comment by ehashman** [comment]
@louisabraham platform-specific shared objects should never go in purelib, how did they get there?

**Pull request merged**